pysb/tools/sensitivity_analysis.py

```diff
@@ -0,0 +1,603 @@
+        return (self.objective_function(function_value)
+                - self._objective_fn_standard) / \
+               self._objective_fn_standard * 100.
+
+    def run(self, save_name=None, out_dir=None):
+        """
+        Run sensitivity analysis
+
+        Parameters
+        ----------
+        save_name : str, optional
+            prefix of saved files
+        out_dir : str, optional
+            location to save output if required
+        """
+        if out_dir is not None:
+            if not os.path.exists(out_dir):
+                os.mkdir(out_dir)
+        if self._solver is None:
+            raise(TypeError, "Must provide a pysb.simulator to run "
+                             "function or when initializing the class")
+        self._solver.initials = None
+        self._solver.param_values = None
+        sim_results = self._solver.run(param_values=None, initials=None)
+        self._objective_fn_standard = self.objective_function(
+            np.array(sim_results.observables[self.observable]))
+
+        traj = self._solver.run(initials=self.simulation_initials)
+        sensitivity_output = np.array(traj.observables)[self.observable].T
+        p_matrix = np.zeros(self._size_of_matrix)
+        p_prime_matrix = np.zeros(self._size_of_matrix)
+        counter = 0
+        # places values in p matrix that are unique
+        for i in range(len(p_matrix)):
+            if i in self.b_index:
+                tmp = self._calculate_objective(sensitivity_output[:, counter])
+                p_matrix[i] = tmp
+                counter += 1
+        # places values in p matrix that are duplicated
+        for i in range(len(p_matrix)):
+            if i in self.b_prime_not_in_b:
+                tmp = self._calculate_objective(
+                    sensitivity_output[:, self.b_prime_not_in_b[i]])
+                p_prime_matrix[i] = tmp
+            elif i in self.b_prime_in_b:
+                p_prime_matrix[i] = p_matrix[self.b_prime_in_b[i]]
+        # Reshape
+        p_matrix = p_matrix.reshape(self._shape_of_matrix)
+        # Project the mirrored image
+        self.p_matrix = p_matrix + p_matrix.T
+        self.p_prime_matrix = p_prime_matrix.reshape(self._shape_of_matrix)
+
+        # save output if desired
+        if save_name is not None:
+            if out_dir is None:
+                out_dir = '.'
+            p_name = os.path.join(out_dir, '{}_p_matrix.csv'.format(save_name))
+            p_prime_name = os.path.join(
+                out_dir, '{}_p_prime_matrix.csv'.format(save_name))
+            print("Saving p matrix and p' matrix to {} and {}".format(
+                p_name, p_prime_name))
+
+            np.savetxt(p_name, self.p_matrix)
+            np.savetxt(p_prime_name, self.p_prime_matrix)
+
+    def _create_index_of_species(self):
+        """ create dictionary of initial conditions by index """
+        index_of_init_condition = {}
+        for i in range(len(self._model.initial_conditions)):
```

Comment:
Iterating over `initial_conditions` and `species` directly with good variable names would make this dense code a lot more readable. I see you do need `j`, so `enumerate` could be used there. Also note that since `initial_conditions` entries are tuples, you can unpack both entries into named variables.